src/shared/copy.c

```diff
@@ -765,6 +765,32 @@ static int memorize_hardlink(
         return 1;
 }
 
+static int prepare_nocow(int fdf, const char *from, int fdt, unsigned *chattr_mask, unsigned *chattr_flags) {
+        unsigned attrs = 0;
+        int r;
+
+        assert(fdf >= 0 || fdf == AT_FDCWD);
+        assert(fdt >= 0);
+        assert(!!chattr_mask == !!chattr_flags);
+
+        if (chattr_mask && FLAGS_SET(*chattr_mask, FS_NOCOW_FL))
+                return 0;
+
+        r = read_attr_at(fdf, from, &attrs);
+        if (r < 0 && !ERRNO_IS_NOT_SUPPORTED(r) && r != -ELOOP)
+                return r;
+
+        if (FLAGS_SET(attrs, FS_NOCOW_FL)) {
+                if (chattr_mask && chattr_flags) {
+                        *chattr_mask |= FS_NOCOW_FL;
+                        *chattr_flags |= FS_NOCOW_FL;
+                } else
+                        (void) chattr_full(fdt, NULL, FS_NOCOW_FL, FS_NOCOW_FL, NULL, NULL, 0);
+        }
```

Comment:
hmm, so, if i set the nocow flag on some dir i then copy a file into with this, we'll end up with the nocow flag enabled with this patch, regardless how the flag was set on the source file.

do we want that? we do, or don't we?

Reply:
"source is COW" is a bit misleading here, because if the source on ext4 then the NOCOW flag will be off, but the source is still not COW at all, given that ext4 doesn#t know such a concept, hence maybe say "source is not NOCOW", just to avoid any confusion